//
// Copyright 2019 The Project Oak Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

syntax = "proto3";

package oak;

// Status values exchanged as i32 values across the Node Wasm interface.
enum OakStatus {
  OAK_STATUS_UNSPECIFIED = 0;
  // Success.
  OK = 1;
  // Invalid handle provided.
  ERR_BAD_HANDLE = 2;
  // Arguments invalid.
  ERR_INVALID_ARGS = 3;
  // Channel has been closed.
  ERR_CHANNEL_CLOSED = 4;
  // Provided buffer was too small for operation (an output value will indicate required size).
  ERR_BUFFER_TOO_SMALL = 5;
  // Argument out of valid range.
  ERR_OUT_OF_RANGE = 6;
}

// Well-known channel handle values.
enum ChannelHandle {
  CHANNEL_HANDLE_UNSPECIFIED = 0;
  LOGGING = 1;   // Write half of channel
  GRPC_IN = 2;   // Read half of channel
  GRPC_OUT = 3;  // Write half of channel
}